### PR TITLE
[SPARK-9572][STREAMING][PYSPARK] Added StreamingContext.getActiveOrCreate() in Python

### PR DESCRIPTION
**Pull request by tdas**

### CODE REVIEW

**Comment by tdas** [comment]
@zsxwing Could you take a quick look?

File: python/pyspark/streaming/context.py [hunk 1]
```diff
@@ -170,6 +173,48 @@ def getOrCreate(cls, checkpointPath, setupFunc):
...
+        but not stopped) or None.
+        """
+        activePythonContext = cls._activeContext
+        if activePythonContext is not None:
+            # Verify that the current running Java StreamingContext is active and is the same one
+            # backing the supposedly active Python context
+            activePythonContextJavaId = activePythonContext._jssc.ssc().hashCode()
+            activeJvmContextOption = activePythonContext._jvm.StreamingContext.getActive()
+            if activeJvmContextOption.isEmpty() or \
+                    activeJvmContextOption.get().hashCode() != activePythonContextJavaId:
```
Comment (zsxwing) [inline python/pyspark/streaming/context.py]: I think it's better to throw an exception when `activeJvmContextOption.get().hashCode() != activePythonContextJavaId` because it will be confusing that `getActive` returns None but `getActiveOrCreate` still complains there is already an active context in Scala codes.

File: python/run-tests.py [hunk 2]
```diff
@@ -158,7 +158,7 @@ def main():
     else:
         log_level = logging.INFO
     logging.basicConfig(stream=sys.stdout, level=log_level, format="%(message)s")
-    LOGGER.info("Running PySpark tests. Output is in python/%s", LOG_FILE)
+    LOGGER.info("Running PySpark tests. Output is in %s", LOG_FILE)
```
Comment (zsxwing) [inline python/run-tests.py]: I think the previous code means the stdout is in `python/unit-tests.log`, which is correct.

File: python/pyspark/streaming/tests.py [hunk 3]
```diff
@@ -1118,4 +1210,4 @@ def search_kinesis_asl_assembly_jar():
                                mqtt_assembly_jar, mqtt_test_jar)
 
     os.environ["PYSPARK_SUBMIT_ARGS"] = "--jars %s pyspark-shell" % jars
-    unittest.main()
+    unittest.main(verbosity=3)
```
Comment (zsxwing) [inline python/pyspark/streaming/tests.py]: unnecessary changes.

File: python/pyspark/streaming/context.py [hunk 4]
```diff
@@ -170,6 +173,48 @@ def getOrCreate(cls, checkpointPath, setupFunc):
...
+                cls._activeContext = None
+        return cls._activeContext
+
+    @classmethod
+    def getActiveOrCreate(cls, checkpointPath, setupFunc):
```
Comment (zsxwing) [inline python/pyspark/streaming/context.py]: Maybe it's better to exchange the locations of `checkpointPath` and `setupFunc`, so that you can use `checkpointPath = None` here.

File: python/run-tests.py (hunk 2 above)
Comment (tdas) [inline python/run-tests.py]: If you see how LOG_FILE is defined, it is already a full path. 
https://github.com/tdas/spark/blob/SPARK-9572/python/run-tests.py#L69

File: python/pyspark/streaming/tests.py (hunk 3 above)
Comment (tdas) [inline python/pyspark/streaming/tests.py]: It is useful for debugging in the generated log file python/unit_tests.log. Without this, the log file only shows which tests have finished, not which ones have started.

File: python/pyspark/streaming/context.py (hunk 1 above)
Comment (tdas) [inline python/pyspark/streaming/context.py]: Yeah, I was on the fence about what to do here. Thanks for deciding for me.

File: python/pyspark/streaming/context.py (hunk 4 above)
Comment (tdas) [inline python/pyspark/streaming/context.py]: The problem is that then this order parameters is opposite of `getOrCreate` which in my own experience is creating a lot confusion when one switches between `getOrCreate` and `getActiveOrCreate`. That's why I explicitly kept the ordering same, as it is more annoying for developers to get the parameters order wrong (more so in Python than Java/Scala) than explicitly provide a checkpoint path or None. What do you think?

File: python/run-tests.py (hunk 2 above)
Comment (zsxwing) [inline python/run-tests.py]: Oh, right. Good catch about the path.

File: python/pyspark/streaming/context.py (hunk 4 above)
Comment (zsxwing) [inline python/pyspark/streaming/context.py]: Good point. Let's keep the current order.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
LGTM.

**Comment by zsxwing** [comment]
retest this please

**Comment by tdas** [comment]
retest this please.

File: python/pyspark/streaming/tests.py [hunk 5]
```diff
@@ -59,12 +60,21 @@ def setUpClass(cls):
...
         self.ssc = StreamingContext(self.sc, self.duration)
 
     def tearDown(self):
-        self.ssc.stop(False)
+        if self.ssc is not None:
+            self.ssc.stop(False)
+        # Clean up in the JVM just in case there has been some issues in Python API
+        jStreamingContextOption = StreamingContext._jvm.SparkContext.getActive()
+        if jStreamingContextOption.nonEmpty():
+            jStreamingContextOption.get().stop(false)
```
Comment (zsxwing) [inline python/pyspark/streaming/tests.py]: `false` -> `False`

Comment (tdas) [inline python/pyspark/streaming/tests.py]: Shouldnt this be `false` based on Java's style of writing *false*

**Comment by zsxwing** [comment]
LGTM
